mitmproxy/protocol/websockets.py

```diff
@@ -0,0 +1,140 @@
+
+        The client starts by sending an Upgrade-request.
+        In order to determine the handshake and negotiate the correct protocol
+        and extensions, the Upgrade-request is forwarded to the server.
+        The response from the server is then parsed and negotiated settings are extracted.
+        Finally the handshake is completed by forwarding the server-response to the client.
+        After that, only WebSockets frames are exchanged.
+
+        PING/PONG frames pass through and must be answered by the other endpoint.
+
+        CLOSE frames are forwarded before this WebSocketsLayer terminates.
+
+        This layer is transparent to any negotiated extensions.
+        This layer is transparent to any negotiated subprotocols.
+        Only raw frames are forwarded to the other endpoint.
+    """
+
+    def __init__(self, ctx, request):
+        super(WebSocketsLayer, self).__init__(ctx)
+        self._request = request
+
+        self.client_key = websockets.get_client_key(self._request.headers)
+        self.client_protocol = websockets.get_protocol(self._request.headers)
+        self.client_extensions = websockets.get_extensions(self._request.headers)
+
+        self.server_accept = None
+        self.server_protocol = None
+        self.server_extensions = None
+
+    def _initiate_server_conn(self):
+        self.establish_server_connection(
+            self._request.host,
+            self._request.port,
+            self._request.scheme,
+        )
+
+        self.server_conn.send(netlib.http.http1.assemble_request(self._request))
+        response = netlib.http.http1.read_response(self.server_conn.rfile, self._request, body_size_limit=None)
+
+        if not websockets.check_handshake(response.headers):
+            raise exceptions.ProtocolException("Establishing WebSockets connection with server failed: {}".format(response.headers))
+
+        self.server_accept = websockets.get_server_accept(response.headers)
+        self.server_protocol = websockets.get_protocol(response.headers)
+        self.server_extensions = websockets.get_extensions(response.headers)
+
+    def _complete_handshake(self):
+        headers = websockets.server_handshake_headers(self.client_key, self.server_protocol, self.server_extensions)
+        self.send_response(models.HTTPResponse(
+            self._request.http_version,
+            101,
+            http.status_codes.RESPONSES.get(101),
+            headers,
+            b"",
+        ))
+
+    def _handle_frame(self, frame, source_conn, other_conn, is_server):
+        self.log(
+            "WebSockets Frame received from {}".format("server" if is_server else "client"),
+            "debug",
+            [repr(frame)]
+        )
+
+        if frame.header.opcode & 0x8 == 0:
+            # forward the data frame to the other side
+            other_conn.send(bytes(frame))
+            # TODO: make some hooks for inline scripts to modify data
```

Comment:
Yes - I already took a look.
However, emitting raw frames is mostly useless for the user since data transfer might be chunked (CONTINUATION) or compressed (gzip/deflate/...)
So this work should probably be done in a separate PR next.